### PR TITLE
Astronomer 1 10 6 experimental knative

### PR DESCRIPTION
**Pull request by dimberman**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: airflow/worker/task_runner_worker.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        # easily exceed the database connection limit when
+        # processing hundreds of simultaneous tasks.
+        settings.configure_orm(disable_connection_pool=True)
+        ti = get_task_instance(dag_id=dag_id,
+                               task_id=task_id,
+                               subdir=process_subdir(subdir),
+                               execution_date=execution_date)
+        out = run(ti)
+        running_tasks_map[key] = ti
+        run_task_instance(ti, log)
```
Comment (ashb) [inline airflow/worker/task_runner_worker.py]: `run(ti)` calls `run_task_instance()`, but we call it again here. Are we running each task twice?

File: airflow/executors/knative_executor.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+            host_header=host_header,
+            pop_running_pipe=child_pop_running_pipe
+        )
+        self.local_loop.start()
+
+    def execute_async(self, key, command, queue=None, executor_config=None, task_instance=None):
+        if queue == "kubernetes":
+            self.kube_executor.execute_async(key, command, queue, executor_config)
+        else:
+            self.task_pipe.send((key, task_instance))
```
Comment (ashb) [inline airflow/executors/knative_executor.py]: I think this is going to arbitrarily fail with "object not pickleable", which is why the SimpleTaskInstance object was created in the first place. Is there anything in particular you need from a TI that isn't on a SimpleTaskInstance object?

Comment (ashb) [reply airflow/executors/knative_executor.py]: If I'm reading the code right this is being used to pull out `(task_id, dag_id, execution_date)`? If so we already have that. `key` is ` tuple of `dag_id, task_id, execution_date, try_number`

**Comment by dimberman** [comment]
```======================================================================
69) FAIL: test_heartbeat_function (tests.worker.test_task_runner_worker.TestTaskRunnerWorker)
----------------------------------------------------------------------
   Traceback (most recent call last):
    /usr/local/lib/python3.6/site-packages/aiohttp/test_utils.py line 478 in new_func
      func(self, *inner_args, **inner_kwargs))
    /usr/local/lib/python3.6/asyncio/base_events.py line 484 in run_until_complete
      return future.result()
    tests/worker/test_task_runner_worker.py line 85 in test_heartbeat_function
      self.assertEqual(stale, [])
   AssertionError: Lists differ: [<TaskInstance: test_requeue_over_dag_conc[77 chars]ng]>] != []
   
   First list contains 1 additional elements.
   First extra element 0:
   <TaskInstance: test_requeue_over_dag_concurrency.test_requeue_over_dag_concurrency_op 2019-01-01 00:00:00+00:00 [running]>
   
   - [<TaskInstance: test_requeue_over_dag_concurrency.test_requeue_over_dag_concurrency_op 2019-01-01 00:00:00+00:00 [running]>]
   + []```

It's really strange... this test is passing just fine on my local breeze environment. I wonder if there is some sort of concurrency issue in travis

**Comment by dimberman** [comment]
Hmm... thinking about it further might have to do with timezones?

**Comment by dimberman** [comment]
```
======================================================================
69) FAIL: test_heartbeat_function (tests.worker.test_task_runner_worker.TestTaskRunnerWorker)
1m----------------------------------------------------------------------
   Traceback (most recent call last):
    /usr/local/lib/python3.6/site-packages/aiohttp/test_utils.py line 478 in new_func
      func(self, *inner_args, **inner_kwargs))
    /usr/local/lib/python3.6/asyncio/base_events.py line 484 in run_until_complete
      return future.result()
    tests/worker/test_task_runner_worker.py line 95 in test_heartbeat_function
      "stale time: {}".format(heartbeat1, heartbeat2, stale_time))
   AssertionError: Lists differ: [<TaskInstance: test_requeue_over_dag_conc[77 chars]ng]>] != []
   
   First list contains 1 additional elements.
   First extra element 0:
   <TaskInstance: test_requeue_over_dag_concurrency.test_requeue_over_dag_concurrency_op 2019-01-01 00:00:00+00:00 [running]>
   
   - [<TaskInstance: test_requeue_over_dag_concurrency.test_requeue_over_dag_concurrency_op 2019-01-01 00:00:00+00:00 [running]>]
   + [] : there's something wrong with the heartbeating.
   Last heartbeat: 2019-11-02 19:37:51.295558+00:00
   current heartbeat: 2019-11-02 19:23:26.414528+00:00
   stale time: 2019-11-02 19:37:59.328217+00:00
```

how is that even possible?

File: tests/worker/test_task_runner_worker.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        dag = DAG(dag_id=dag_id, start_date=DEFAULT_DATE,
+                  max_active_runs=1, concurrency=2)
+        task = DummyOperator(task_id=task_id, dag=dag)
+        import datetime
+        start_date = datetime.datetime(year=2019, day=1, month=1)
+        ti = TI(task=task, execution_date=start_date, state=State.RUNNING)
+
+        task_runner_worker.running_tasks_map["test_requeue_over_dag_concurrency_op"] = ti
+        await asyncio.sleep(5)
+        text = await resp.text()
```
Comment (dimberman) [inline tests/worker/test_task_runner_worker.py]: Ok I think what's happening is that the heartbeating from this test is spilling over into the next test. This of course leads to a larger question of why the heartbeat in the second test isn't overriding this old test. Will check further.

**Comment by dimberman** [comment]
There must be something up with the heartbeat function. I added an extra test to ensure that there's no leftover stale tasks when the test starts. It seems pretty consistently 15 minutes behind ```   - [<TaskInstance: test_requeue_over_dag_concurrency.test_requeue_over_dag_concurrency_op 2019-01-01 00:00:00+00:00 [running]>]
   + [] : there's something wrong with the heartbeating.
   Last heartbeat: 2019-11-03 15:46:00.092676+00:00
   current heartbeat: 2019-11-03 15:31:14.419225+00:00
   stale time: 2019-11-03 15:46:08.124166+00:00```

**Comment by dimberman** [comment]
Interesting. When I ONLY run those two tests on travis it seems that they actually pass. I'm gonna start putting a bit more time between to see if I can suss out more

**Comment by dimberman** [comment]
https://travis-ci.com/astronomer/airflow/builds/134764521 yeah so here's an example where when I ONLY run those two tests it passes but when put into the whole test gauntlet it fails.

**Comment by dimberman** [comment]
`======================================================================
69) FAIL: test_heartbeat_function (tests.worker.test_task_runner_worker.TestTaskRunnerWorker)
----------------------------------------------------------------------
   Traceback (most recent call last):
    /usr/local/lib/python3.6/site-packages/aiohttp/test_utils.py line 478 in new_func
      func(self, *inner_args, **inner_kwargs))
    /usr/local/lib/python3.6/asyncio/base_events.py line 484 in run_until_complete
      return future.result()
    tests/worker/test_task_runner_worker.py line 98 in test_heartbeat_function
      "stale time: {}".format(heartbeat1, heartbeat2, stale_time))
   AssertionError: Lists differ: [<TaskInstance: test_heartbeat_func.test_h[49 chars]ng]>] != []
   
   First list contains 1 additional elements.
   First extra element 0:
   <TaskInstance: test_heartbeat_func.test_heartbeat_func_op 2019-01-01 00:00:00+00:00 [running]>
   
   - [<TaskInstance: test_heartbeat_func.test_heartbeat_func_op 2019-01-01 00:00:00+00:00 [running]>]
   + [] : there's something wrong with the heartbeating.
   Last heartbeat: 2019-11-04 00:58:26.812646+00:00
   current heartbeat: 2019-11-04 00:43:39.332394+00:00
   stale time: 2019-11-04 00:58:41.840253+00:00`

Ok so a bit of good news. I changed the name of the task to a name unique to JUST that function, and it appears that it still has the EXACT same 15 minute issue. This means that the issue isn't some value being stale but there is in fact something funky either about how we're heartbeating or how the heartbeating is working in asyncio.

**Comment by dimberman** [comment]
Ok the test also fails locally when run with all of the other tests...

**Pull request closed without merging**